core/src/main/java/com/datastax/oss/driver/api/core/AsyncPagingIterable.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core;
+
+import com.datastax.oss.driver.api.core.cql.AsyncResultSet;
+import com.datastax.oss.driver.api.core.cql.ExecutionInfo;
+import com.datastax.oss.driver.internal.core.AsyncPagingIterableWrapper;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Iterator;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+/**
+ * An iterable of elements which are fetched asynchronously by the driver, possibly in multiple
+ * requests.
+ */
+public interface AsyncPagingIterable<ElementT> {
```

Comment:
There *could* be a self type here, because of `fetchNextPage()`. I tried to go down that route, but it wasn't pretty; it propagates to the result of `map`, but it can't be materialized on the client side:
```java
PagingIterable<?, Integer> iterable = resultSet.map(row -> row.getInt(0));
```
Also, the implementation of `AsyncPagingIterableWrapper` was horrendous.

So I chose to go the simpler route and override the method with a covariant type in `ResultSet`.

Reply:
I wonder if a self type for `fetchNextPage` wouldn't even be conceptually wrong, because the contract of this method does not mandate that implementors should return a future of themselves (iow, some implementation A might return a future that once completed yields an object of type B).

That said, I would like to I understand what was so horrible regarding the implementation of `AsyncPagingIterableWrapper`; [I gave it a try myself](https://github.com/datastax/java-driver/commit/270a2473b27dc191e1fcdcd0b23daa7e1f581b52) and found it actually pretty manageable.

Reply:
I ended up having to parameterize it with 4 types: the source and target iterables, and the source and target element types, and there were still unchecked casts.
Indeed your implementation looks simpler. I noticed your `fetchNextPage` doesn't return a covariant future, not sure if this has an impact or not. Maybe I just over-complicated things.

But since we agree that a self-type is not desirable (the `?` in the caller code is the worst aspect IMO), I won't dig into it further.